src/3d/qgsvirtualpointcloudentity_p.cpp

```diff
@@ -0,0 +1,201 @@
+  }
+
+  updateBboxEntity();
+  connect( provider(), &QgsVirtualPointCloudProvider::subIndexLoaded, this, &QgsVirtualPointCloudEntity::createChunkedEntityForSubIndex );
+}
+
+QList<QgsChunkedEntity *> QgsVirtualPointCloudEntity::chunkedEntities() const
+{
+  return mChunkedEntitiesMap.values();
+}
+
+QgsVirtualPointCloudProvider *QgsVirtualPointCloudEntity::provider() const
+{
+  return qobject_cast<QgsVirtualPointCloudProvider *>( mLayer->dataProvider() );
+}
+
+QgsAABB QgsVirtualPointCloudEntity::boundingBox( int i ) const
+{
+  return mBboxes.at( i );
+}
+
+void QgsVirtualPointCloudEntity::createChunkedEntityForSubIndex( int i )
+{
+  const QVector<QgsPointCloudSubIndex> subIndexes = provider()->subIndexes();
+  QgsPointCloudLayerChunkedEntity *newChunkedEntity = new QgsPointCloudLayerChunkedEntity( subIndexes.at( i ).index(),
+      mMap,
+      mCoordinateTransform,
+      static_cast< QgsPointCloud3DSymbol * >( mSymbol->clone() ),
+      mMaximumScreenSpaceError,
+      mShowBoundingBoxes,
+      mZValueScale,
+      mZValueOffset,
+      mPointBudget );
+
+  mChunkedEntitiesMap.insert( i, newChunkedEntity );
+  newChunkedEntity->setParent( this );
+  connect( newChunkedEntity, &QgsChunkedEntity::pendingJobsCountChanged, this, &Qgs3DMapSceneEntity::pendingJobsCountChanged );
+  emit newEntityCreated( newChunkedEntity );
+}
+
+void QgsVirtualPointCloudEntity::handleSceneUpdate( const QgsChunkedEntity::SceneState &state )
+{
+  const QVector<QgsPointCloudSubIndex> subIndexes = provider()->subIndexes();
+  for ( int i = 0; i < subIndexes.size(); ++i )
+  {
+    const QgsAABB &bbox = mBboxes.at( i );
+
+    if ( bbox.isEmpty() )
+      continue;
+
+    // magic number 256 is the common span value for a COPC root node
+    constexpr int SPAN = 256;
+    const float epsilon = std::min( bbox.xExtent(), bbox.yExtent() ) / SPAN;
+    const float distance = bbox.distanceFromPoint( state.cameraPos );
+    const float sse = Qgs3DUtils::screenSpaceError( epsilon, distance, state.screenSizePx, state.cameraFov );
+    constexpr float THRESHOLD = .2;
+    const bool displayAsBbox = sse < THRESHOLD;
+    if ( !displayAsBbox && !subIndexes.at( i ).index() )
+      provider()->loadSubIndex( i );
+
+    setRenderSubIndexAsBbox( i, displayAsBbox );
+    if ( !displayAsBbox && mChunkedEntitiesMap.contains( i ) )
+      mChunkedEntitiesMap[i]->handleSceneUpdate( state );
+  }
+  updateBboxEntity();
+}
+
+QgsRange<float> QgsVirtualPointCloudEntity::getNearFarPlaneRange( const QMatrix4x4 &viewMatrix ) const
+{
+  float fnear = 1e9;
+  float ffar = 0;
+
+  for ( QgsChunkedEntity *entity : mChunkedEntitiesMap )
+  {
+    if ( entity->isEnabled() )
+    {
+      const QgsRange<float> range = entity->getNearFarPlaneRange( viewMatrix );
+      ffar = std::max( range.upper(), ffar );
+      fnear = std::min( range.lower(), fnear );
+    }
+  }
+  return QgsRange<float>( fnear, ffar );
```

Comment:
Usually the terrain is the one that would give a valid range, similar to what the bboxes would give. But if there's no terrain, or some raster terrain smaller than the scene extent, then we would end up with bad values, indeed. Let's use the bboxes as a fallback in case we have no chunked entities enabled.